### ISSUE TITLE
Mismatch between shape of tensors (due to vocabulairy size)

### ISSUE THREAD

**Issue by ZheMann** [rkfg/gpt-2#3]
First, I call `!sh scripts/createspmodel.sh combined_wo_startendtag.txt 8192` to generate `hparams.json`, `sp.model` and `sp.vocab` (with a vocabulairy size of 8192). 

Then, when trying to train my gpt-2 model, I get the following error:
```
InvalidArgumentError (see above for traceback): Assign requires shapes of both tensors to match. lhs shape= [8192,768] rhs shape= [50257,768]
	 [[node save/Assign_147 (defined at ./train.py:139) ]]
```
This is caused in train.py by line 139:
```
        saver = tf.train.Saver(
            var_list=all_vars,
            max_to_keep=5,
            keep_checkpoint_every_n_hours=2)
```
and line 155: `saver.restore(sess, ckpt)`

My question is: do you have any idea where rhs shape = [50257,768] comes from? I know the LHS comes from my hparams.json, and when I generate sp.model and sp.vocab on a vocabulairy size of 50257, everyting works fine. However, I need a smaller size currently. 

I was wondering whether you also ran into this problem and if so, how you fixed it.

Many thanks in advance.

**Comment by ZheMann** [comment]
@rkfg do you have any idea?

**Comment by rkfg** [comment]
That script creates `hparams.json` in the directory from which you call it. Copy it together with `sp.model` and `sp.vocab` to `/models/<modelname>`. This shape probably comes from the original model by OpenAI. Sorry, for some reason I don't receive issue notifications for new issues (if I'm not directly mentioned), I see there's another one created already.

**Comment by ZheMann** [comment]
> This shape probably comes from the original model by OpenAI. 

I think so too, but I was wondering whether you also ran into this issue while training gpt-2 with the custom generated `hparams.json`, `sp.model` and `sp.vocab` for the Russian language and. Otherwise I will use the vocabulairy size from the original model (50527) as this does work.

**Comment by rkfg** [comment]
No, this should not happen. Did you create a separate empty directory for your model? You're not supposed to tweak the original GPT-2 as it uses a different dictionary and only knows English.

**Comment by ZheMann** [comment]
I do as follows:

1. Run `!python3 download_model.py 117M` to download the model into directory `gpt-2/models/117M`
2. Run `!sh scripts/createspmodel.sh combined_wo_startendtag.txt 8192` which creates `hparams.json`, `sp.model` and `sp.vocab` in directory gpt-2. 
3. Copy `hparams.json`, `sp.model` and `sp.vocab`  to `gpt-2/models/117M`
4. Run `!PYTHONPATH=src ./train.py --dataset combined_wo_startendtag.txt --model_name '117M' --save_every 10000 --restore_from 'latest'`

I'm not sure why I would have to create a separate empty directory for my model, but probably that's the cause of this problem? 

Thanks for your time, anyways.

**Comment by rkfg** [comment]
You probably misunderstood the point of this repository. I created these scripts to train the model from scratch without any prior data. To start you only need the hyperparameters file (`hparams.json`) and the dictionary files, that's all! After you start training the script will generate and rotate checkpoints. You can't "restore from latest" checkpoint because there's no checkpoint for your model at all. You can't load the pretrained model for another dictionary using your own dictionary (of a different size and content) and expect it to somehow magically start working with it. The network accepts and produces mere numbers from 0 to <dictionary_size>, in case of GPT-2 it would be 50527. Each number is substituted with a token, several letters, that altogether produce words and sentences. When you create your own dictionary those numbers no longer match the same characters, the pretrained model becomes completely useless. And even if the dictionary sizes match it won't work as you expect because the dictionary creation algorithm (and the data to derive it from) is different from what OpenAI guys used so it's no different than starting from the blank slate. Even worse probably because the model already created some internal "rules" to produce text and you feed it completely illogical garbage instead, re-learning will take much more time than starting from scratch.

**Comment by ZheMann** [comment]
I feel pretty stupid after your comment. I already wondered how the hell this could work for the pre-trained gpt-2 model, but after your comment it all makes sense. This pre-trained model will obivously not work with different dictionary files (what was I thinking). 

So by using this repository I am able to create my own gpt-2 (-ish) model from scratch. Thank you so much for taking the time to write such a detailed explanation, and to be honest, this makes this repository even more interesting. Although it might take some time to train a decent model from scratch, I am going to try it for the Dutch language.

**Comment by rkfg** [comment]
You're welcome! It definitely takes a lot of time and I'm not sure if I was doing it right because the loss value almost doesn't change after about 4 days of training (and the text quality is not that good while still kinda interesting). I have no idea if I hit the floor and that's the best I could do with these parameters or if I should just leave it training further. I also suspect that loss does not truly reflects the network's ability to produce interesting texts. Probably it might still improve while the loss doesn't change much. Since the cross-entropy loss function is logarithmic by nature it's possible that decreasing loss by 0.01 from 3.7 is a significant improvement compared to the same 0.01 from 4.5 for example.

Another improvement might be using tensor units in the specialized GPUs like V100 but from what I read it would require using 16-bit floating points and this particular network is fully 32 bit. There are some parameters in TensorFlow that allow for automatic conversion back and forth but either I didn't do it right or the improvement was negligible. I did training on my GTX 1070 and also tried Amazon's GPU instances but it was not worth it at all. Barely 2x improvement for the money I could run my own GPU for a month (and on Amazon I spent it for a mere hour!). So if you achieve any improvement in that area be sure to hit me up!

**Comment by ZheMann** [comment]
> Probably it might still improve while the loss doesn't change much

Thanks for letting me know. I'm looking forward experimenting with it and I'm excited to see the results after training the model for several days. 

About using tensor units: I just read the following [issue](https://github.com/minimaxir/gpt-2-simple/issues/3) on minimaxir's fork of gpt-2, where someone was able to run gpt-2 on TPU in [colab](https://colab.research.google.com/drive/17I7VZrcxM-BfadRqAFWb2DzG3RoAtG5n#scrollTo=rO9_HdZnJ0xg). Although some minor bugs seem to appear, it looks interesting and this is definitely something I could look further into.
